### PR TITLE
DateRange: new component

### PR DESCRIPTION
**Pull request by AlbertCarreras**
#### What changed?

DateRange: new component
Screenshot by Dropbox Capture

#### Why?

New component
### Links

- [Jira](https://jira.pinadmin.com/browse/GESTALT-5268)
- [Docs](https://coda.io/d/Gestalt-X-Team-Collaboration_dqIC9SDPsU9/DateRange-WIP_suvq1?utm_source=slack#_luq5a)
- [Figma](https://www.figma.com/file/SnFbvwSV6IeU1B2H69MbKn/Documentation-Date-range-component?type=design&node-id=651-67457&mode=design&t=KQwjWDWGqgvMzmDy-0)

### Checklist

- [x] Added unit and Flow Tests
- [X] Added documentation + accessibility tests
- [X] Verified accessibility: keyboard & screen reader interaction
- [X] Checked dark mode, responsiveness, and right-to-left support
- [X] Checked stakeholder feedback (e.g. Gestalt designers)

### CODE REVIEW

File: packages/gestalt-datepicker/src/DateRange/InternalDatePicker.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            ?.date({ width: 'short' })
+            .replace(/(y{1,4})/, 'yyyy')
+            .replace(/(d{1,2})/, 'dd')
+            .replace(/(M{1,2})/, 'MM') ?? 'MM/dd/yyyy',
+        );
+      }
+    }, [localeData]);
+
+    return (
+      <div className="_gestalt">
```
Comment (rlingineni) [inline packages/gestalt-datepicker/src/DateRange/InternalDatePicker.js:65]: Can we give this a min/max-height? It's kind of jarring when the height changes when the calendar is moved from left to right. 

I think it should always be 6 rows to be safe:

image

**Review by rlingineni — CHANGES_REQUESTED** [review]
wip: review

File: docs/graphics/general/DateRange.svg
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply docs/graphics/general/DateRange.svg:1]: Yes

File: packages/gestalt-datepicker/src/DateRange.js [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+  const { applyText, cancelText } = useDefaultLabel('DateRange');
+
+  return (
+    <Box rounding={4} color="default" borderStyle="shadow" minHeight={425} display="inlineBlock">
+      <Flex>
+        {radioGroup ? (
+          <div className={borderStyles.borderRight}>
+            <Box paddingY={4} paddingX={6} width={216}>
+              {radioGroup}
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:173]: Flow checks for it

File: packages/gestalt-datepicker/src/DateRange.js [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+                minDate={minDate}
+                maxDate={maxDate}
+              />
+            </Box>
+            <Flex.Item alignSelf="end">
+              <Flex gap={2}>
+                <Box marginBottom={4} marginEnd={4}>
+                  <Button color="transparent" text={cancelText} onClick={() => onCancel()} />
+                </Box>
+                <Box marginBottom={4} marginEnd={4}>
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:243]: Question for Layla, I follow designs

File: packages/gestalt-datepicker/src/DateRange/InternalDatePicker.js (hunk 1 above)
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange/InternalDatePicker.js:65]: Discussed already with designers. Decided to leave as is.

File: packages/gestalt-datepicker/src/DateField/InternalDateField.js [hunk 4]
```diff
@@ -100,7 +100,7 @@ const CustomTextField = forwardRef(
           onKeyDown={onKeyDown}
           onMouseUp={onMouseUp}
         />
-        {!disabled && !readOnly ? (
+        {!disabled && !readOnly && ownerState?.passthroughProps?.onClearInput ? (
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateField/InternalDateField.js:103]: Internal UI Datefield internal prop when customizing input field

File: packages/gestalt-datepicker/src/DateRange.js [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+              {radioGroup}
+            </Box>
+          </div>
+        ) : null}
+        <Box>
+          <Flex alignItems="start" justifyContent="center" direction="column">
+            <div className={borderStyles.dateFieldSection}>
+              <Flex gap={3}>
+                <Box width={280}>
+                  <InternalDateField
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:182]: Locale data

File: docs/pages/visual-test/DateRange-light.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+// @flow strict
+import { type Node } from 'react';
+import { Box, ColorSchemeProvider } from 'gestalt';
+import { DateRange } from 'gestalt-datepicker';
+
+export default function Snapshot(): Node {
+  return (
+    <ColorSchemeProvider colorScheme="light">
+      <Box color="default" padding={2} width={675}>
+        <DateRange
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply docs/pages/visual-test/DateRange-light.js:10]: Can you elaborate?

File: docs/graphics/general/DateRange.svg
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply docs/graphics/general/DateRange.svg:1]: We need Kate to produce several

File: packages/gestalt-datepicker/src/DateRange.js [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+                    onError={onStartDateError}
+                    onBlur={onStartDateBlur}
+                    onFocus={onStartDateFocus}
+                    value={startDateValue}
+                    minDate={minDate}
+                    maxDate={maxDate}
+                    errorMessage={startDateErrorMessage}
+                  />
+                </Box>
+                <Box dangerouslySetInlineStyle={{ __style: { marginTop: '15px' } }}>
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:199]: I'll look into it, but we did this things. For you to consider in your proposal. Pay off the complexity. We do these adjustments here and there

File: packages/gestalt-datepicker/src/DateRange.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (rlingineni) [reply packages/gestalt-datepicker/src/DateRange.js:243]: @laylaleeal

File: packages/gestalt-datepicker/src/DateRange.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rlingineni) [reply packages/gestalt-datepicker/src/DateRange.js:173]: Yes, but awe ran into an issue with Callout. One of the customers was using a $flow-fix-me and passing in arbitrary content and it was a pain to fix.

We do it in SlimBanner
https://github.com/pinterest/gestalt/blob/ad1750ffd668bfcf023c4a11ba73616d562bc38c/packages/gestalt/src/SlimBanner.js#L253

File: docs/pages/visual-test/DateRange-light.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (rlingineni) [reply docs/pages/visual-test/DateRange-light.js:10]: There's no values selected. In the dark/light-mode examples, can you select values so we can see those colors as well

File: packages/gestalt-datepicker/src/DateRange.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (laylaleeal) [reply packages/gestalt-datepicker/src/DateRange.js:243]: The buttons should be a ButtonGroup, which has a set spacing between Buttons. We don't want to change that! and about the primary color, we had this discussion, but decided to go with a Secondary/Tertiary pair to avoid a case of having too many primary buttons in an interface (there's no wash applied when DateRange is open, so the page underneath is still visible).

[the remark this answers is not in the document]
Comment (rlingineni) [reply packages/gestalt-datepicker/src/DateRange.js:243]: Cool!

Also, tiny feedback on the "Apply" language, which seems to be for a filter. Like, if I click apply I expect an action to happen immediately, e.g apply a filter, but in the case if it's used as part of larger form,  apply may not do immediate results. should we call it "done" like [Flights](https://www.google.com/travel/flights), and [Expedia](https://www.expedia.com/)

image

File: packages/gestalt-datepicker/dist/index.d.ts [hunk 8]
```diff
@@ -420,6 +500,11 @@ export interface DateFieldProps {
  * =========================================================
  */
 
+/**
+ * https://gestalt.pinterest.systems/web/dateranger
```
Comment (rlingineni) [inline packages/gestalt-datepicker/dist/index.d.ts:504]: typo

File: packages/gestalt-datepicker/src/DateRange.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js]: FIXED, DONE

File: packages/gestalt-datepicker/src/DateRange.test.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+// @flow strict-local
+import { create } from 'react-test-renderer';
+import DateRange from './DateRange.js';
+
+describe('DateRange', () => {
+  it('renders', () => {
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.test.js]: DONE

File: packages/gestalt-datepicker/src/DatePicker.css [hunk 10]
```diff
@@ -21,6 +21,16 @@
   position: relative;
 }
 
+.react-datepicker-inline {
+  background-color: var(--g-colorGray50);
+  border-radius: 16px;
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DatePicker.css:26]: not for border-radius, padding margin top DONE

File: packages/gestalt-datepicker/src/DateRange.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:187]: Got it

File: packages/gestalt-datepicker/src/DateRange.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:262]: @laylaleeal Apply or Done?

File: packages/gestalt-datepicker/dist/index.d.ts (hunk 8 above)
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/dist/index.d.ts:504]: I wanted to be the 6th power ranger, DateRanger

Comment (AlbertCarreras) [reply packages/gestalt-datepicker/dist/index.d.ts:504]: DONE

File: packages/gestalt-datepicker/src/DateRange.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (laylaleeal) [reply packages/gestalt-datepicker/src/DateRange.js:264]: sorry, just seeing this - I'm ok with Done! cc @hectoid any thoughts on it?

[the remark this answers is not in the document]
Comment (hectoid) [reply packages/gestalt-datepicker/src/DateRange.js:264]: @laylaleeal and @rlingineni I understand that travel sites use "Done", but the primary use case for us is Ads and Analytics. Our competitors use "Apply" or "Update". We currently use "Apply" elsewhere in the Pinterest Business interface. I think we should stick to that and can update if things change.
image

image

image

[the remark this answers is not in the document]
Comment (hectoid) [reply packages/gestalt-datepicker/src/DateRange.js:264]: But, now I also think we should have a primary / secondary button treatment to match other dropdowns. If that is still a doable change @AlbertCarreras 
image

[the remark this answers is not in the document]
Comment (hectoid) [reply packages/gestalt-datepicker/src/DateRange.js:264]: (also notice they're not following our Sentence case guidelines for buttons though)

[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:264]: @laylaleeal @hectoid can we find consensus on the button treatment? https://github.com/pinterest/gestalt/pull/3001#discussion_r1265753701

File: packages/gestalt/src/README_DELETE_DateRange.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
+# DateRange
+
+FOLLOW STEPS HERE
+https://gestalt.pinterest.systems/get_started/developers/contributing/development_process#Create-a-pull-request
+
+These are the remaining steps to configure your components:
```
Comment (rlingineni) [inline packages/gestalt/src/README_DELETE_DateRange.md:6]: Reminder to remove file before merge

File: packages/gestalt-datepicker/src/DateRange.js [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+  /**
+   * Minimal selectable date. Disables any date values before the provided date.  See the [disable future and past variant](https://gestalt.pinterest.systems/web/datefield#Disable-future-and-past) to learn more.
+   */
+  minDate?: Date,
+  /**
+   * Callback triggered when the user clicks the Cancel button to not persist the selected dates. It should be used to close DateRange. See the [controlled component variant](https://gestalt.pinterest.systems/web/daterange#Controlled-component) to learn more.
+   */
+  onCancel: () => void,
+  /**
+   * Callback triggered when the user blurs the input of the end date DateField. See the [error messaging variant](https://gestalt.pinterest.systems/web/daterange#Error-messaging) to learn more.
```
Comment (rlingineni) [inline packages/gestalt-datepicker/src/DateRange.js:77]: Maybe rephrase to: `Callback triggered when the end date input loses focus`, since I feel like a user doesn't proactively blur

File: packages/gestalt-datepicker/src/DateRange.js [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+  |}) => void,
+  /**
+   * Callback triggered when the user focus on the input of the end date DateField. See the [error messaging variant](https://gestalt.pinterest.systems/web/daterange#Error-messaging) to learn more.
+   */
+  onEndDateFocus?: ({|
+    event: SyntheticFocusEvent<HTMLInputElement>,
+    value: string,
+  |}) => void,
+  /**
+   * Callback triggered when the user blurs the input of the start date DateField. See the [error messaging variant](https://gestalt.pinterest.systems/web/daterange#Error-messaging) to learn more.
```
Comment (rlingineni) [inline packages/gestalt-datepicker/src/DateRange.js:102]: same as above

File: packages/gestalt-datepicker/src/DateRange.js [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+                  onStartDateChange({ value: startDate });
+                  onEndDateChange({ value: endDate });
+                }}
+                minDate={minDate}
+                maxDate={maxDate}
+              />
+            </Box>
+            <Flex.Item alignSelf={isMobile ? 'center' : 'end'}>
+              <ButtonGroup>
+                <Box marginBottom={4} marginEnd={4}>
```
Comment (rlingineni) [inline packages/gestalt-datepicker/src/DateRange.js:260]: We may not need to wrap in Box anymore in the ButtonGroup since it's adding extra spacing?

File: docs/examples/daterange/errorMessaging.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
+// @flow strict
+import { type Node, useState } from 'react';
+import { DateRange } from 'gestalt-datepicker';
+
+export default function Example(): Node {
+  const [startDate, setStartDate] = useState<Date | null>(null);
```
Comment (rlingineni) [inline docs/examples/daterange/errorMessaging.js:6]: Can we initialize with an invalid date so the error appears by default

File: docs/pages/web/daterange.js [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+          <MainSection.Card
+            cardSize="md"
+            sandpackExample={
+              <SandpackExample code={errorMessaging} name="error example" previewHeight={500} />
+            }
+          />
+        </MainSection.Subsection>
+        <MainSection.Subsection
+          title="Disable future & past"
+          description="DateField supports disabling future and past dates from being selected."
```
Comment (rlingineni) [inline docs/pages/web/daterange.js:228]: How? What props need to be set?

File: docs/pages/web/daterange.js [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+            title="Disable past"
+            sandpackExample={
+              <SandpackExample code={disabledPast} name="past example" previewHeight={500} />
+            }
+          />
+          <MainSection.Card
+            cardSize="md"
+            title="Disable future"
+            sandpackExample={
+              <SandpackExample code={disabledFuture} name="future example" previewHeight={500} />
```
Comment (rlingineni) [inline docs/pages/web/daterange.js:241]: I think disabled past and future can be done in the same example with just a fixed range of dates

File: docs/pages/web/daterange.js [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+            }
+          />
+        </MainSection.Subsection>
+        <MainSection.Subsection
+          title="Supporting locales"
+          description="Select the correct locale to see DataRange's selected locale support."
```
Comment (rlingineni) [inline docs/pages/web/daterange.js:247]: *Daterange

I think could be rephrased as:

> Daterange supports multiple locales. Use the dropdown to try out different locales by passing in the `localeData` prop.

If it's going to be in the variant section, I think it might be helpful to have a code example

File: docs/pages/web/daterange.js [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+        </MainSection.Subsection>
+      </MainSection>
+
+      <MainSection name="Best practices">
+        <MainSection.Subsection columns={2}>
+          <MainSection.Card
+            type="do"
+            title="Do"
+            description={`
+- Disable future or past dates according to the use case. If the user would like to see predictions, for example, disable the past.
```
Comment (rlingineni) [inline docs/pages/web/daterange.js:152]: Do we need examples here?

File: docs/pages/web/daterange.js [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+`}
+          />
+        </MainSection.Subsection>
+      </MainSection>
+
+      <MainSection name="Variants">
+        <MainSection.Subsection
+          title="Controlled component"
+          description={`
+DateRange is a controlled component. Follow the implementation in the example to implement a controlled DateRange correctly.
```
Comment (rlingineni) [inline docs/pages/web/daterange.js:171]: nit: I think we should this[ exact same format](https://gestalt.pinterest.systems/web/datefield#Controlled-component), and list out the props

File: docs/examples/daterange/errorMessaging.js [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+  const minDate = new Date(year, 6, 1);
+  const maxDate = new Date(year, 6, 31);
+
+  return (
+    <DateRange
+      endDateValue={endDate}
+      endDateErrorMessage={endErrorMessage}
+      minDate={minDate}
+      maxDate={maxDate}
+      onStartDateBlur={() => {
```
Comment (rlingineni) [inline docs/examples/daterange/errorMessaging.js:29]: Just wondering, any reason we aren't handling error message internally? I'm not sure I follow what the blur is doing? Seems like a lot of handlers have to be defined externally just to get it to show

File: packages/gestalt-datepicker/src/DateRange.js [hunk 22]
```diff
@@ -0,0 +1,285 @@
...
+                    onError={onStartDateError}
+                    onBlur={onStartDateBlur}
+                    onFocus={onStartDateFocus}
+                    value={startDateValue}
+                    minDate={minDate}
+                    maxDate={maxDate}
+                    errorMessage={startDateErrorMessage}
+                  />
+                </Box>
+                <Box display="flex" height="100%" alignItems="center">
```
Comment (rlingineni) [inline packages/gestalt-datepicker/src/DateRange.js:215]: In an error state, this messes up

image

File: docs/pages/web/daterange.js [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+        >
+          <MainSection.Card
+            cardSize="md"
+            sandpackExample={
+              <SandpackExample code={errorMessaging} name="error example" previewHeight={500} />
+            }
+          />
+        </MainSection.Subsection>
+        <MainSection.Subsection
+          title="Disable future & past"
```
Comment (rlingineni) [inline docs/pages/web/daterange.js:227]: nit: maybe rephrase as  
>Disable past and future dates

**Review by rlingineni — COMMENTED** [review]
No major feedback

File: packages/gestalt-datepicker/src/DateRange.js (hunk 22 above)
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:215]: haha that's why I added a marginTop! @rlingineni

File: docs/pages/web/daterange.js (hunk 23 above)
Comment (AlbertCarreras) [reply docs/pages/web/daterange.js:227]: DONE

File: packages/gestalt/src/README_DELETE_DateRange.md (hunk 11 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/README_DELETE_DateRange.md:6]: DONE

File: packages/gestalt-datepicker/src/DateRange.js (hunk 12 above)
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:77]: DONE

File: packages/gestalt-datepicker/src/DateRange.js (hunk 13 above)
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:102]: DONE

File: docs/examples/daterange/errorMessaging.js (hunk 15 above)
Comment (AlbertCarreras) [reply docs/examples/daterange/errorMessaging.js:6]: DONE

File: packages/gestalt-datepicker/src/DateRange.js (hunk 14 above)
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:260]: DONE

File: docs/pages/web/daterange.js (hunk 16 above)
Comment (AlbertCarreras) [reply docs/pages/web/daterange.js:228]: DONE

File: docs/pages/web/daterange.js (hunk 17 above)
Comment (AlbertCarreras) [reply docs/pages/web/daterange.js:241]: I'm going to keep them separately has there are different use cases for future (select new campaign dates for activation) to past (select previous campaign dates for analytics)

File: docs/pages/web/daterange.js (hunk 18 above)
Comment (AlbertCarreras) [reply docs/pages/web/daterange.js:247]: DONE first part. 

If we decide to implement date-fns in the codesandboxes sure. If I do example, it would break atm.

We can do that as a separate project for all cmps using data-fns

File: docs/pages/web/daterange.js (hunk 19 above)
Comment (AlbertCarreras) [reply docs/pages/web/daterange.js:152]: We decided not to

File: docs/pages/web/daterange.js (hunk 20 above)
Comment (AlbertCarreras) [reply docs/pages/web/daterange.js:171]: DONE

File: packages/gestalt-datepicker/src/DateRange.js (hunk 22 above)
Comment (AlbertCarreras) [reply packages/gestalt-datepicker/src/DateRange.js:215]: DONE

File: docs/examples/daterange/errorMessaging.js (hunk 21 above)
Comment (AlbertCarreras) [reply docs/examples/daterange/errorMessaging.js:35]: DONE

File: docs/pages/web/datepicker.js [hunk 24]
```diff
@@ -221,7 +221,7 @@ export default function DatePickerPage({ generatedDocGen }: {| generatedDocGen:
           description="DatePicker supports disabling future & past dates as well as an array of selected dates."
         >
           <MainSection.Card
-            title="Disable future & past"
+            title="Disable past & future dates"
```
Comment (rlingineni) [inline docs/pages/web/datepicker.js:224]: Heads up, noticing these weird extra headers:
image

If we use a bold header/title like this, can you add a description underneath

**Comment by rlingineni** [comment]
Can we center the calendar here in the localizaton example, or make it take full width?
image

File: docs/pages/web/daterange.js [hunk 25]
```diff
@@ -0,0 +1,339 @@
...
+          />
+        </MainSection.Subsection>
+      </MainSection>
+
+      <MainSection name="Variants">
+        <MainSection.Subsection
+          title="Controlled component"
+          description={`
+DateRange is a controlled component.
+Use \`endDateValue\`,  \`startDateValue\`,  \`onEndDateChange, \`onStartDateChange\`, \`onEndDateError\`, \`onStartDateError\`, \`onSubmit\`and \`onCancel\` to implement it correctly.
```
Comment (rlingineni) [inline docs/pages/web/daterange.js:172]: This formatting is incorrect

image

Comment (AlbertCarreras) [reply docs/pages/web/daterange.js:172]: DONE

File: docs/pages/web/datepicker.js (hunk 24 above)
Comment (AlbertCarreras) [reply docs/pages/web/datepicker.js:224]: DONE
